doc/man-source/lirc-lsremotes.1
.TH lirc-lsremotes "1" "Last change: Aug 2015" "lirc-lsremotes @version@" "User Commands"
.SH NAME
.P
\fBlirc-lsremotes\fR - parse lirc configuration files.
.SH SYNOPSIS
.P
\fBlirc-lsremotes\fR [\fIoptions\fR] <directory>
.P
\fBlirc-lsremotes\fR [\fIoptions\fR] <file>


.SH DESCRIPTION
lirc-lsremotes can parse lirc remote configuration files (i. e., lircd.conf)
and display a condensed index line, the complete data and/or diagnostics.
.P
When given a directory (first form) it expects this to contain a number of
vendor directories, each of which containing one or more configuration files.
.P
In the second form lirc-lsremotes displays remote configurations for a given
config file.

.SH OPTIONS
.TP 4
.B -d, --dump
Dump the complete configuration data for listed configurations.
.TP 4
.B  -s, --silent
Only print diagnostics.
.TP 4
.B -v --version
Print version info
.TP 4
.B -h, --help
Print help message.
.SH EXAMPLES
Print a status line and possible diagnostics for a file:
.IP "" 4
.nf
$ lirc-lsremotes  remotes/hitachi/RM-466E.lircd.conf
.br
lirc-lsremotes: Warning RM-466E: you should specify a valid gap value
.br
hitachi;RM-466E.lircd.conf;no_lircmd;no_photo;RM-466E;timing;no_raw
.fi
.P
List a status line for all remotes under remotes, saving diagnostics into
a file:
.IP "" 4
.nf
$ lirc-lsremotes  remotes > remotes.list 2>warnings.txt
.fi
.

.SH "SEE ALSO"
http://sourceforge.net/projects/lirc-remotes


NEWS
0.9.4

  * The select() calls in lircd, lircrcd and library routines have
    been replaced with poll() to avoid crashes when file descriptors
    becomes > 1024 (#109).
  * The documentation is heavily revised. Old obsolete text is
    removed, the manual is reorganized, some info is moved to
    manpages and a lot of minor updates and bugfixes.
  * The remotes/ directory is not used any more, removed.
  * Set supplementary groups using effective-user (RH bz 1253907).
  * Work around kernel bug in /dev/lirc0 output (#119).

0.9.3pre2 1/08/15
  * irtoy: New support for debug LEDs. New documentation
    file, support in lirc-setup and driver table doc added.
  * The library routines for parsing lircd.conf now warns for
    duplicate keys and values (#107).
  * lirc-lsremotes now displays logs after parsing files, making
    more parsing warnings visible.
  * lirc-config-tool is fixed to comply with guidelines (#105).
  * Bug in irexec when using modes fixed (#102).
  * Options parsing (notably -U) and just listing first remote
    bugs in irsimsend have been fixed.
  * A new kernel driver irpipe has been added. This allows feeding data
    from userspace to lircd, which can use the /dev/irpipeX device
    as a clean substitute for /dev/lircX. A companion CLI tool
    irpipe(1) is also added.
0
0.9.3pre1 30/06/15

  * The irexec semantics have been changed so that the invoked
    programs run disconnected from the main irexec process.
    See irexec(1) (#92).
  * mode2 has been updated to handle also LIRCCODE devices (#94)
    and the --driver-options switch
  * xmode2 has been updated to handle udp driver, the new
    --driver-options switch + some general overhaul.
  * irrecord has got some love:
     + It handles multiple definitions of same symbol OK (#78)
     + Warns about single-key config files (#77)
     + Checks for ambient light disturbances (#76)
     + New option allows updating of existing template file.
     + Template handling bug fixed
     + -l --list-namespace option parsing bug fixed (#101).
  * irrecord has been split into a library and a cli interface
    (#36). The library interface is unstable.
  * lircmd: Several option parsing bugs fixed, one of which
    blocked the systemd service (#104, RH bz 1234155).
  * lirc-setup: fixed name clash on recent python versions.
  * The udp driver is patched to allow adjustable clockrates,
    see new udp_driver.html driver doc.
  * The mplay driver has been updated with patches which has
    been available.
  * Overall code formatting is improved with more consistent
    layout and updated git hook. See CONTRIBUTE.md
  * The compiler has been switched to c++, and filenames are
    updated to *.cpp (#79)
  * New build dependency: xsltproc (usually coming with libxml).
  * Unlinked and hard to find driver docs and manpages have
    been linked.
  * Build: now uses parallell make by default.
  * Docs: The Configuration Guide has been updated.

0.9.2 12/09/14
  * Documentation fixes

0.9.2pre3  12/04/14
  * Add fix for #70, by mistake omitted in pre2.

0.9.2pre2  11/30/14
  * Fixes for #67 - #75 besides #71 and #74.

0.9.2pre1  11/21/14
  * Removed functionality:
    + The smode2 program has been retired from the build, although it still
      is available in contrib/.
    + The setup.sh script is removed (build always builds all drivers).
  * Drivers are loaded dynamically. Many programs including lircd now
    needs to know where the drivers are located to be able to use
    them. There are new options, environment variables and hard-coded
    defaults. See the manpages for more.
  * A new tool lirc-lsplugins is added to inspect dynamic driver modules.
  * New, unstable configuration script lirc-setup. The script guides user
    when selecting both capture device and remote config. It sports auto-
    detection of most devices, searching and downloading from the remotes
    database and a considerable amount of bugs.
  * The remotes/ directory has become obsolete, only used for testing.
    New configuration tools uses the remotes database at
    http://sf.net/p/lirc-remotes as the only source (#53). A new tool
    irdb-get(1) supports searching and downloading from remotes database.
  * Building on MacOSX has been refreshed.
  * New devices and features:
    + Support for the Dangerous Prototypes irtoy device is added (#20).
    + Support for for sonyir on MacOSX is added. This is a single hiddev
      device; with the framework in place hopefully more devices will be
      added over time.
    + Support for the DIY assembly kit yard2 usb ir device is added. This is
      an experimental driver which depends on a server which currently is not
      packaged in any major distro; a Debian/Ubuntu PPA is available.
    + Support for the Roku remote protocol added (#8).
    + Support for long keypresses has been added. See the "Configuring LIRC"
      chapter in the manual, search for "ignore_first_event" (#50).
  * lircd has been verified to run as regular user as long as it can open
    devices and such. A new option --effective-user makes lircd drop it's
    privileges after opening devices if started as root. A new udev rule
    is shipped to make /dev/lirc* accessible for users in group 'lirc" (#56).
  * Logging runtime option is renamed: debug -> loglevel. Levels can be given
    symbolically e. g., 'info' or 'error'. Old option name is still supported.
  * Configuration changes
    + The default location for the user lircrc-file has been changed from
      ~/.lircrc to ~/.config/lircrc. The old location is still used as fallback.
      (ticket #52).
    + The lircd.conf configuration has been updated. lircd will now besides
      the lircd.conf file also read all *.conf files in the
      /etc/lirc/lircd.conf.d directory. The already existing include directive
      has been modified to accept wildcards and also been documented.
    + lircd.conf has a new 'driver' attribute for configs requiring a specific

tools/lirc-lsremotes.cpp
/****************************************************************************
** lirc-lsremotes **********************************************************
****************************************************************************
*
* lirc-lsremotes - list remotes from the remotes database.
*
*/

#include <config.h>

#include <dirent.h>
#include <getopt.h>
#include <errno.h>
#include <fnmatch.h>
#include <glob.h>
#include <libgen.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <unistd.h>

#include "lirc_private.h"
#include "lirc_client.h"

static int opt_silent = 0;
static int opt_dump = 0;

static const char* current_dir = NULL;


static const char* const USAGE =
	"List remotes from a copy of the remotes database.\n"
	"Synopsis:\n"
	"    lirc-lsremotes <path> [remote]\n"
	"    lirc-lsremotes [-h | -v]\n\n"
	"<path> is the path to the remotes directory or a single file.\n"
	"[remote] is remotes to list, using wildcards. Defaults  to '*'.\n\n"
	"Options:\n"
	"    -s  --silent       Just parse and print diagnostics.\n"
	"    -d  --dump         Dump complete configuration (noisy).\n"
	"    -v, --version      Print version.\n"
	"    -h, --help         Print this message.\n";


static struct option options[] = {
	{ "dump",    no_argument, NULL, 'd' },
	{ "silent",  no_argument, NULL, 's' },
	{ "help",    no_argument, NULL, 'h' },
	{ "version", no_argument, NULL, 'v' },
	{ 0,	     0,		  0,	0   }
};

/** Given lircd.conf path, write lircmd path or "no_lircmd" in buff. */
void get_lircmd(const char* path, char* buff, ssize_t size)
{
	char pattern[256];
	char* base;
	char* dir;
	char* ext;

	strncpy(buff, path, size - 1);
	ext = strstr(buff, ".lircd.conf");
	if (ext != NULL)
		*ext = '\0';
	base = basename(buff);
	dir = dirname(buff);
	strncpy(pattern, dir, sizeof(pattern) - 1);
	strncat(pattern, "/", sizeof(pattern) - strlen(pattern) - 1);
	strncat(pattern, base, sizeof(pattern) - strlen(pattern) - 1);
	strncat(pattern, ".lircmd.conf", sizeof(pattern) - strlen(pattern) - 1);
	if (access(pattern, R_OK) == 0)
		strncpy(buff, basename(pattern), size - 1);
	else
		strcpy(buff, "no_lircmd");
}


/** Given lircd.conf path, write photo path or "no_photo" in buff. */
void get_photo(const char* path, char* buff, ssize_t size)
{
	glob_t globbuf;
	int r;
	char* ext;

	strncpy(buff, path, size);
	ext = strstr(buff, ".lircd.conf");
	if (ext != NULL)
		*ext = '\0';
	strncat(buff,
		"{.jpg,.png,.gif,.JPG,.PNG,.GIF}",
		size - strlen(buff) - 1);
	r = glob(buff, GLOB_BRACE, NULL, &globbuf);
	if (r == GLOB_NOMATCH) {
		strcpy(buff, "no_photo");
		return;
	}
	if (globbuf.gl_pathc > 1)
		logprintf(LIRC_WARNING, "Multiple photos for %s", buff);
	strncpy(buff, basename(globbuf.gl_pathv[0]), size);
}


/** Print a line for each remote definition in lircd.conf file on path. */
void print_remotes(const char* path)
{
	char my_path[256];
	char photo[256];
	char lircmd[256];
	struct ir_remote* r;
	FILE* f;
	const char* dir;
	const char* base;
	const char* timing;

	strncpy(my_path, path, sizeof(my_path));
	base = basename(my_path);
	dir = dirname(my_path);
	if (strrchr(dir, '/') != NULL)
		dir = strrchr(dir, '/') + 1;
	f = fopen(path, "r");
	if (f == NULL) {
		fprintf(stderr, "Cannot open %s (!)\n", path);
		return;
	}
	r = read_config(f, path);
	if (opt_silent)
		return;
	while (r != NULL && r != (void*)-1) {
		timing = r->pzero != 0 || r->pzero != 0 || is_raw(r) ?
			 "timing" : "no_timing";
		strncpy(photo, path, sizeof(photo));
		get_photo(path, photo, sizeof(photo));
		get_lircmd(path, lircmd, sizeof(lircmd));
		printf("%s;%s;%s;%s;%s;%s;%s;%s\n",
		       dir,
		       base,
		       lircmd,
		       photo,
		       r->name,
		       timing,
		       is_raw(r) ? "raw" : "no_raw",
		       r->driver != NULL ? r->driver : "no_driver");
		fflush(stdout);
		if (opt_dump)
			fprint_remote(stdout, r, "Dumped by lirc-lsremotes");
		r = r->next;
	}
	;
	fclose(f);
}


int isdir(const struct dirent* ent)
{
	struct stat statbuf;
	char buff[256];

	snprintf(buff, sizeof(buff), "%s/%s", current_dir, ent->d_name);
	if (stat(buff, &statbuf) == -1)
		return 0;
	return S_ISDIR(statbuf.st_mode);
}


/** Return true if argument defines a lircd.conf file. */
int isfile(const struct dirent* ent)
{
	const char* dot = strrchr(ent->d_name, '.');

	if (dot != NULL) {
		if (strcasecmp(dot + 1, "jpg") == 0)
			return 0;
		if (strcasecmp(dot + 1, "gif") == 0)
			return 0;
		if (strcasecmp(dot + 1, "png") == 0)
			return 0;
		if (strcasecmp(dot + 1, "html") == 0)
			return 0;
		if (strcasecmp(dot + 1, "txt") == 0)
			return 0;
	}
	if (fnmatch("*lircrc*", ent->d_name, 0) == 0)
		return 0;
	return !isdir(ent);
}


/** List all remotes found in files in dir. */
void listdir(const char* dirname)
{
	char dirpath[256];
	char filepath[256];
	struct dirent** namelist;
	int size;
	int i;

	if (strcmp(dirname, "..") == 0)
		return;
	if (strlen(dirname) > 0 && dirname[0] == '.')
		return;
	snprintf(dirpath, sizeof(dirpath), "%s/%s", current_dir, dirname);
	size = scandir(dirpath, &namelist, isfile, alphasort);
	for (i = 0; i < size; i += 1) {
		snprintf(filepath, sizeof(filepath), "%s/%s",
			 dirpath, namelist[i]->d_name);
		free(namelist[i]);
		print_remotes(filepath);
	}
}


int lsremotes(const char* dirpath, const char* remote)
{
	struct dirent** namelist;
	int size;
	int i;
	struct stat statbuf;

	if (stat(dirpath, &statbuf) == -1)
		return 2;
	if (!S_ISDIR(statbuf.st_mode)) {
		print_remotes(dirpath);
		return 0;
	}
	current_dir = dirpath;
	size = scandir(dirpath, &namelist, isdir, alphasort);
	for (i = 0; i < size; i += 1) {
		listdir(namelist[i]->d_name);
		free(namelist[i]);
	}
	return 0;
}


int check_logs(const char* path)
{
	struct stat statbuf;
	FILE* f;
	char buff[256];

	if (stat(path, &statbuf) == -1)
		return 1;
	if (statbuf.st_size == 0)
		return 1;
	f = fopen(path, "r");
	if (f == NULL) {
		fprintf(stderr, "Cannot open logfile %s for read", path);
		return 0;
	}
	while (fgets(buff, sizeof(buff), f) != NULL)
		fprintf(stderr, buff);
	fclose(f);
	return 0;
}

int main(int argc, char** argv)
{
	const char* configs;
	const char* dirpath;
	char path[128];
	int c;

	while ((c = getopt_long(argc, argv, "shdv", options, NULL)) != EOF) {
		switch (c) {
		case 'd':
			opt_dump = 1;
			break;
		case 's':
			opt_silent = 1;
			break;
		case 'h':
			puts(USAGE);
			return EXIT_SUCCESS;
		case 'v':
			printf("%s\n", "lirc-lsremotes " VERSION);
			return EXIT_SUCCESS;
		case '?':
			fprintf(stderr, "unrecognized option: -%c\n", optopt);
			fputs("Try `lirc-lsremotes -h' for more information.\n",
			      stderr);
			return EXIT_FAILURE;
		}
	}
	if (argc == optind + 2) {
		dirpath = argv[optind];
		configs = argv[optind + 1];
	} else if (argc == optind + 1) {
		dirpath = argv[optind];
		configs = "*";
	} else {
		fputs(USAGE, stderr);
		return EXIT_FAILURE;
	}
	lirc_log_get_clientlog("lirc-lsremotes", path, sizeof(path));
	unlink(path);
	lirc_log_set_file(path);
	lirc_log_open("lirc-lsremotes", 1, LIRC_NOTICE);
	lsremotes(dirpath, configs);
	check_logs(path);
	return 0;
}
